jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/objects/MetadataAsmFactory.java

```diff
@@ -372,14 +376,28 @@ public AnnotationVisitor visitArray(String name) {
 
         @Override
         public void visitEnd() {
-            if (this.element != null) {
-                if (this.annotation.isMeta()) {
-                    this.element.addMetaAnnotation(this.annotation);
+            if ( this.element != null && !isRecursive()  ) {
+                if ( this.annotation.isMeta() ) {
+                    this.element.addMetaAnnotation( this.annotation );
                 } else {
-                    this.element.addAnnotation(this.annotation);
+                    this.element.addAnnotation( this.annotation );
                 }
             }
         }
+        private boolean isRecursive() {
+            if(this.element == null || this.annotation == null)
```

Comment:
_this.element == null_ is already being checked before this method is called
null value of _this.annotation_ would cause NPE in this.annotation.isMeta() so this is expected always != null